src/Features/CSharp/Portable/UseRecursivePatterns/AnalyzedNodeExpander.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseRecursivePatterns
+{
+    using static AnalyzedNode;
+
+    // We replace pattern variables with their original input in order to find and unify identical nodes.
+    // For instance:
+    //
+    //     `e.x.y is var v && v is p`
+    //
+    // Becomes:
+    //
+    //     `e.x.y is var v && e.x.y is p`
+    //
+    // Which will be in turn reduced to:
+    //
+    //     `e is { x: { y: p } }`
+    //
+    // TODO: Unused pattern variables should be elided in a later pass.
+    // TODO: Can we use an annotation + reducers for this purpose?
```

Comment:
@CyrusNajmabadi Is it conventional/possible to use reducers for this? It would really simplify things since we don't have to deal with any specific case inside the refactoring.

The idea is attach an annotation to designators and look for a diagnostic (unnecessary assignment) to remove it in a separate reducer. I'll have to investigate to see it's workable.

Reply:
That's not possible because available SemanticModel doesn't support diagnostics or data flow.